src/vptranslator.rs

```diff
@@ -0,0 +1,253 @@
+}
+///Allows you to present a list of dates in a readable way.
+pub fn translate(dates: &BTreeSet<Date>) -> BlockPattern {
+    let mut res = BlockPattern::default();
+    let start_date = match dates.iter().next() {
+        Some(d) => *d,
+        None => return res,
+    };
+    let end_date = match dates.iter().next_back() {
+        Some(d) => *d,
+        None => return res,
+    };
+
+    let validity_period = ValidityPeriod {
+        start_date: start_date,
+        end_date: end_date,
+    };
+    res.validity_periods.push(validity_period);
+
+    let mut monday_ref = get_prev_monday(start_date);
+    let validity_pattern = compute_validity_pattern(monday_ref, end_date, &dates);
+    let best_week = get_min_week_pattern(&validity_pattern);
+    fill_operating_days(best_week, &mut res.operating_days);
+
+    for week in validity_pattern {
+        if week != best_week {
+            let exception: u8 = (!best_week) & week;
+            fill_exceptions(
+                monday_ref,
+                exception,
+                ExceptionType::Add,
+                &mut res.exceptions,
+            );
+
+            let exception: u8 = (week ^ best_week) & best_week;
+            fill_exceptions(
+                monday_ref,
+                exception,
+                ExceptionType::Remove,
+                &mut res.exceptions,
+            );
+        }
+        monday_ref += Duration::days(7);
+    }
+    clean_extra_dates(start_date, end_date, &mut res.exceptions);
+    res
+}
+
+//       July 2012
+// Mo Tu We Th Fr Sa Su
+//                    1
+//  2  3  4  5  6  7  8
+//  9 10 11 12 13 14 15
+// 16 17 18 19 20 21 22
+// 23 24 25 26 27 28 29
+// 30 31
+
+#[test]
+fn nb_weeks() {
+    let mut dates = BTreeSet::new();
+    let mut start_date: Date;
+    let mut end_date: Date;
+
+    start_date = Date::from_ymd(2012, 7, 2);
+    assert_eq!(get_prev_monday(start_date), start_date);
+
+    start_date = Date::from_ymd(2012, 7, 5);
+    assert_eq!(get_prev_monday(start_date), Date::from_ymd(2012, 7, 2));
+
+    start_date = Date::from_ymd(2012, 7, 8);
+    assert_eq!(get_prev_monday(start_date), Date::from_ymd(2012, 7, 2));
+
+    // one week
+    start_date = Date::from_ymd(2012, 7, 2);
+    end_date = Date::from_ymd(2012, 7, 8);
+    dates.insert(start_date);
+    dates.insert(end_date);
+    assert_eq!(
+        compute_validity_pattern(get_prev_monday(start_date), end_date, &dates).len(),
+        1
+    );
+
+    // partial + one week
+    dates.clear();
+    start_date = Date::from_ymd(2012, 7, 4);
+    end_date = Date::from_ymd(2012, 7, 15);
+    dates.insert(start_date);
+    dates.insert(end_date);
+    assert_eq!(
+        compute_validity_pattern(get_prev_monday(start_date), end_date, &dates).len(),
+        2
+    );
+
+    // one week + partial
+    dates.clear();
+    start_date = Date::from_ymd(2012, 7, 2);
+    end_date = Date::from_ymd(2012, 7, 13);
+    dates.insert(start_date);
```

Comment:
to ease the writting of the test, you can use https://github.com/bluss/maplit
that makes it possible to write:
`let dates = btreeset!{Date::from_ymd(2012, 7, 2), Date::from_ymd(2012, 7, 13)};`

Reply:
but start_date is re-use after

Reply:
since lots of tests seems to share the same patterns, in addition to split them you can maybe add a test utility function to reduce the noise.

to add a test utility function you can do:

```rust

#[cfg(test)]
mod tests {
   fn the_utility_function() {... }

    #[test]
    fn one_function(){...}
}
```


the utility could be 
`compute_vp_between(date1, date2) -> Vec<u8>`